src/awkward/_util.py

```diff
@@ -44,124 +72,357 @@ def is_sized_iterable(obj):
+            except Exception:
+                return x._repr("    ", "", "")
+
+        elif isinstance(x, ak.record.Record):
+            try:
+                return str(ak.highlevel.Record(x))
+            except Exception:
+                return x._repr("    ", "", "")
+
+        else:
+            return repr(x)
+
+
+def error(exception, error_context=None):
+    if isinstance(exception, type) and issubclass(exception, Exception):
+        try:
+            exception = exception()
+        except Exception:
+            return exception
+
+    if isinstance(exception, (NotImplementedError, AssertionError)):
+        return type(exception)(
+            str(exception)
+            + "\n\nSee if this has been reported at https://github.com/scikit-hep/awkward-1.0/issues"
+        )
+
+    if error_context is None:
+        error_context = ErrorContext.primary()
+
+    if isinstance(error_context, ErrorContext):
+        # Note: returns an error for the caller to raise!
+        return type(exception)(error_context.format_exception(exception))
     else:
-        date = " (target date: " + date + ")"
-    warning = """In version {}{}, this will be {}.
+        # Note: returns an error for the caller to raise!
+        return exception
 
-To raise these warnings as errors (and get stack traces to find out where they're called), run
 
-    import warnings
-    warnings.filterwarnings("error", module="awkward.*")
+def indexerror(subarray, slicer, details=None):
+    detailsstr = ""
+    if details is not None:
+        detailsstr = f"""
 
-after the first `import awkward` or use `@pytest.mark.filterwarnings("error:::awkward.*")` in pytest.
+Error details: {details}."""
 
-Issue: {}.""".format(
-        version, date, will_be, message
-    )
-    warnings.warn(warning, category, stacklevel=stacklevel + 1)
+    error_context = ErrorContext.primary()
+    if not isinstance(error_context, SlicingErrorContext):
+        # Note: returns an error for the caller to raise!
+        return IndexError(
+            f"cannot slice {type(subarray).__name__} with {SlicingErrorContext.format_slice(slicer)}{detailsstr}"
+        )
+
+    else:
+        # Note: returns an error for the caller to raise!
+        return IndexError(
+            error_context.format_exception(
+                f"at inner {type(subarray).__name__} of length {subarray.length}, using sub-slice {error_context.format_slice(slicer)}.{detailsstr}"
+            )
+        )
 
 
-# Sentinel object for catching pass-through values
-class MISSING:
-    pass
+###############################################################################
 
+# # Enable warnings for the Awkward package
+# warnings.filterwarnings("default", module="awkward.*")
 
-virtualtypes = (ak.layout.VirtualArray,)
 
-unknowntypes = (ak.layout.EmptyArray,)
+# def deprecate(
+#     message,
+#     version,
+#     date=None,
+#     will_be="an error",
+#     category=DeprecationWarning,
+#     stacklevel=2,
+# ):
```

Comment:
Nevermind, here it is: it's just commented out in v2 _util.py (and GitHub is making a mess of the difference).

So it just needs to be un-commented.